executor/aggregate.go

```diff
@@ -395,32 +395,48 @@ func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, f
 }
 
 // getGroupKey evaluates the group items and args of aggregate functions.
-func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
-	w.groupValDatums = w.groupValDatums[:0]
+func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, input *chunk.Chunk) (err error) {
+	numRows := input.NumRows()
+	if len(w.groupKey) < numRows {
+		for i := len(w.groupKey); i < numRows; i++ {
+			w.groupKey = append(w.groupKey, make([]byte, 10*len(w.groupByItems)))
+		}
+	}
+	for i := 0; i < numRows; i++ {
+		w.groupKey[i] = w.groupKey[i][:0]
+	}
 	for _, item := range w.groupByItems {
-		v, err := item.Eval(row)
+		tp := item.GetType()
+		buf, err := expression.GetColumn(tp.EvalType(), numRows)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		// This check is used to avoid error during the execution of `EncodeDecimal`.
-		if item.GetType().Tp == mysql.TypeNewDecimal {
-			v.SetLength(0)
+		defer expression.PutColumn(buf)
```

Comment:
This is wired, `buf` may not be put back except the last one. We remove `defer`.